### PR TITLE
add service security

### PR DESCRIPTION
**Pull request by andrei-astaschenko**

### CODE REVIEW

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+@ConditionalOnClass({ Feign.class })
+@ConditionalOnProperty(value = "feign.oauth2.enabled", matchIfMissing = true)
+public class OAuth2FeignAutoConfiguration {
+
+    @Bean
+    @ConditionalOnBean(OAuth2ClientContext.class)
+    public RequestInterceptor oauth2FeignRequestInterceptor(OAuth2ClientContext oauth2ClientContext) {
+        return new OAuth2FeignRequestInterceptor(oauth2ClientContext);
+    }
+}
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java]: new line

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * Creates new instance of {@link OAuth2FeignRequestInterceptor} with client context.
+     *
+     * @param oauth2ClientContext the OAuth2 client context
+     */
+    public OAuth2FeignRequestInterceptor(OAuth2ClientContext oauth2ClientContext) {
+        Assert.notNull(oauth2ClientContext, "Context can not be null");
+        this.oauth2ClientContext = oauth2ClientContext;
+    }
+
+    /**
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: no needs

File: src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    @Bean
+    @Qualifier("tokenStore")
+    protected TokenStore tokenStore() {
+        return new JwtTokenStore(jwtAccessTokenConverter);
+    }
+
+    @Bean
+    protected JwtAccessTokenConverter jwtTokenEnhancer() {
+        JwtAccessTokenConverter converter =  new JwtAccessTokenConverter();
+        Resource resource = new ClassPathResource("public.cert");
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:32]: Is this piece of code verified by running JAR file?

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+     * The logger instance used by this class.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(OAuth2FeignRequestInterceptor.class);
+
+    /**
+     * The authorization header name.
+     */
+    private static final String AUTHORIZATION_HEADER = "Authorization";
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: AUTHORIZATION_HEADER_KEY

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java (hunk 2 above)
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: TOKEN_TYPE

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * The {@code Bearer} token type.
+     */
+    private static final String BEARER_TOKEN_TYPE = "Bearer";
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: TOKEN_TYPE

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        this.oauth2ClientContext = oauth2ClientContext;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void apply(RequestTemplate template) {
+        if (template.headers().containsKey(AUTHORIZATION_HEADER)) {
+            LOGGER.warn("The Authorization token has been already set");
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: Why do we log message as a warning?

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        } else if (oauth2ClientContext.getAccessTokenRequest().getExistingToken() == null) {
+            LOGGER.warn("Can not obtain existing token for request, if it is a non secured request, ignore.");
+        } else {
+            LOGGER.debug("Constructing Header {} for Token {}", AUTHORIZATION_HEADER, BEARER_TOKEN_TYPE);
+            template.header(AUTHORIZATION_HEADER, String.format("%s %s", BEARER_TOKEN_TYPE,
+                    oauth2ClientContext.getAccessTokenRequest().getExistingToken().toString()));
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: the toString method should be called aiutomatically, shouldn't it?

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        Assert.notNull(oauth2ClientContext, "Context can not be null");
+        this.oauth2ClientContext = oauth2ClientContext;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void apply(RequestTemplate template) {
+        if (template.headers().containsKey(AUTHORIZATION_HEADER)) {
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: It looks like we can ignore first 2 cases (no needs in if-clauses)

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java (hunk 4 above)
Comment (Andrei-Khadziukou) [reply src/main/java/com/epam/petclinic/order/config/OAuth2FeignRequestInterceptor.java]: Could you please also check whether there is a constant in Spring with this value?

File: src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        String publicKey = null;
+        try {
+            publicKey = new String(FileCopyUtils.copyToByteArray(resource.getInputStream()));
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java]: What will be responsible for closing stream?

File: src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java (hunk 3 above)
Comment (andrei-astaschenko) [reply src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:32]: It's verified.

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java (hunk 1 above)
Comment (andrei-astaschenko) [reply src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java]: Line is added.

File: src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import feign.RequestInterceptor;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.oauth2.client.OAuth2ClientContext;
+
+@Configuration
+@ConditionalOnClass({ Feign.class })
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/order/config/OAuth2FeignAutoConfiguration.java]: If my understanding is correct we will get compile exception if there is no Feign class in classpath (we need in to compile this class according to import clause)?

**Comment by shalauniou** [comment]
Could you please add to `readme.md` or `Poc` doc how to test it and how it works, and add java docs for configuration classes what they responsible for?

File: src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+
+    @Bean
+    @Qualifier("tokenStore")
+    protected TokenStore tokenStore() {
+        return new JwtTokenStore(jwtAccessTokenConverter);
+    }
+
+    @Bean
+    protected JwtAccessTokenConverter jwtTokenEnhancer() {
```
Comment (shalauniou) [inline src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:30]: Do we really need defining JwtAccessTokenConverter as a `@Bean`? Is it used outside JwtConfiguration?

Comment (andrei-astaschenko) [reply src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:30]: Yes, Bean is requiered to convert jwt token.

Comment (shalauniou) [reply src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:30]: If I rewrite it as `return new JwtTokenStore(jwtTokenEnhancer() )`   - why it won't be work?

Comment (andrei-astaschenko) [reply src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:30]: In this case, We get invalid message from service which can not convert jwt token.

Comment (shalauniou) [reply src/main/java/com/epam/petclinic/order/config/JwtConfiguration.java:30]: Ok

**Pull request merged**